### DISCUSSION TITLE
Compatibility Issues with Line Breaks in Chinese Fonts

### DISCUSSION THREAD

**Discussion by ILearn-better** [pymupdf/PyMuPDF#2622] [Q&A] [answered]
_**Please provide all mandatory information!**_

## Describe the bug (mandatory)
I want to use `insert_textbox` to insert Chinese character but the result style is different from  my expectation. 
## To Reproduce (mandatory)
        #文本翻译与插入
```python
        for block in page_text_message:
            rate = 6
            x0, y0, x1, y1 = [round(i / sampling) for i in block["bbox"]]
            coordinate = (x0 - rate, y0 - rate, x1 + rate, y1 + rate)
            text = page.get_textbox(coordinate)
            # text = keep_last_line_break(text)
            print("the text I get:{}".format(text))
            translated_text = interface_to_dict(text)["data"]
            print("translated result:{}".format(translated_text))
            print("====" * 8)

            x0, y0, x1, y1 = [round(i) for i in block["bbox"]]
            coordinate = (x0, y0, x1, y1)
            font_path = os.path.join(os.getcwd(),"font","arial-unicode-ms.ttf")
            ff = new_pdf_page.insert_font(fontname="F0",
                                          fontfile=font_path,
                                          fontbuffer=None,
                                          set_simple=False)
            # 字体推荐:https://github.com/Haixing-Hu/latex-chinese-fonts
            if not isinstance(translated_text, str):
                translated_text = str(translated_text).replace("\n", " ")
            new_pdf_page.insert_textbox(coordinate, translated_text,
                                        fontsize=9 * sampling, lineheight=1, fontname="F0", fill=None, overlay=True)
```
To avoid being too lengthy in presentation,I just  show a part of my code.
I guess the reason why it turned out is because of the compatibility of fonts to `insert_box` . 
## Expected behavior (optional)
Aleady I have `replace("\n "," ")`，but when I insert the translated text it will still occur line break,which isn't that I want it happen. 
## Screenshots (optional)
Expected:
This photo from html to pdf.
![图片](https://github.com/pymupdf/PyMuPDF/assets/65653618/4c36fc72-43ec-4477-a454-479490284f77)
Don't expect:
This photo from class `fitz`.
![图片](https://github.com/pymupdf/PyMuPDF/assets/65653618/9b4473f0-3601-42aa-81aa-93f5e9ff4cd3)

## Your configuration (mandatory)

[GCC 7.3.0] 
 linux 
 
PyMuPDF 1.20.2: Python bindings for the MuPDF 1.20.3 library.
Version date: 2022-08-13 00:00:01.
Built for Python 3.8 on linux (64-bit).

**Comment by JorjMcKie** [comment]
This is no bug but a Discussions item. Let me transfer first.

**Comment by JorjMcKie** [comment, marked as answer]
The logic to break text into pieces that fit in the given rectangle width, roughly does this (as far as I remember):

* Split given text in lines (`\n`).
* Then, everything is handled as a word that is surrounded by spaces.
* In each line, the next word is inserted as long as it does not lap outside the right border. If it does, start a new line.
* If we are at the left text box border and the word is longer than the available width, forcedly break the word into pieces that do fit.

Please note that **_only standard spaces_** `0x20` are handled this way.
Also, if your text does not contain a comma follwed by a standard space like this `", "` but instead some Chinese character looking similar, then no line break will take place. Similar with other punctuation.
I hope you see the point.

Also, you must be using a different front from what is shown in the HTML output.

**Reply by ILearn-better** [reply]
Thanks for your reply.

**Reply by ihnfsax** [reply]
Can PyMuPDF treat a single Chinese character as a word?

**Reply by JorjMcKie** [reply]
No. A word must be surrounded by spaces or line break characters.

**Comment by ILearn-better** [comment]
## Recording
There is an another question.When the area covered by the text I input exceeds the area of the box, using `insert_textbox` becomes ineffective, resulting in the content I intended to insert not appearing on the PDF.

![图片](https://github.com/pymupdf/PyMuPDF/assets/65653618/e5ece544-add5-4b2a-baec-601c217863ab)
============================================================================================
![图片](https://github.com/pymupdf/PyMuPDF/assets/65653618/194fe4ee-f40c-4ba3-8fa1-2b8843715996)

**Comment by JorjMcKie** [comment]
What is your question? I only read a statement: "... becomes ineffective."

**Reply by ILearn-better** [reply]
In the second picture,I input a long long string in three text area,but just one area show the string.
I found if the length of the text is out of the box,`insert_box` will output nothing,but I hope it can truncate text and output.

**Reply by JorjMcKie** [reply]
The documentation also says that the method has a return code, which is a negative float if the text did not fit in the box! Did you read this?
The value of that float is the number of points that are missing in the text box's height. So just increase the text box rectangle by at least this value ... and you are good!
